### ISSUE TITLE
Planner chooses 6-15 times longer parallel seq scan instead of index scan on ivfflat.

### ISSUE THREAD

**Issue by pashkinelfe** [pgvector/pgvector#187]
When experimenting with dbpedia-openai-1M base and ivfflat index (lists = 2000, probes = 50) I've noticed that default scan plan is seqscan and is 6-15 itmes longer than the index scan: 

SET ivfflat.probes = 50;
EXPLAIN (ANALYZE, BUFFERS) select id from emb order by vector <#> (select vector from emb_query order by random() limit 1) limit 10;

Default plan (max_parallel_workers = 0):
```
 Limit  (cost=19763.14..19764.31 rows=10 width=12) (actual time=6214.830..6215.003 rows=10 loops=1)
   Buffers: shared hit=6634013 read=2
   InitPlan 1 (returns $0)
     ->  Limit  (cost=239.00..239.00 rows=1 width=40) (actual time=2.664..2.666 rows=1 loops=1)
           Buffers: shared hit=64
           ->  Sort  (cost=239.00..264.00 rows=10000 width=40) (actual time=2.664..2.665 rows=1 loops=1)
                 Sort Key: (random())
                 Sort Method: top-N heapsort  Memory: 25kB
                 Buffers: shared hit=64
                 ->  Seq Scan on emb_query  (cost=0.00..189.00 rows=10000 width=40) (actual time=0.021..1.
344 rows=10000 loops=1)
                       Buffers: shared hit=64
   ->  Gather Merge  (cost=19524.14..107030.25 rows=750000 width=12) (actual time=6214.829..6214.998 rows=
10 loops=1)
         Workers Planned: 2
         Params Evaluated: $0
         Workers Launched: 0
         Buffers: shared hit=6634013 read=2
         ->  Sort  (cost=18524.12..19461.62 rows=375000 width=12) (actual time=6211.728..6211.729 rows=10
loops=1)
               Sort Key: ((emb.vector <#> $0))
               Sort Method: top-N heapsort  Memory: 25kB
               Buffers: shared hit=6633949 read=2
               ->  Parallel Seq Scan on emb  (cost=0.00..10420.50 rows=375000 width=12) (actual time=1.200
..6088.076 rows=900000 loops=1)
                     Buffers: shared hit=6633949 read=2
 Planning:
   Buffers: shared hit=113
 Planning Time: 0.548 ms
 Execution Time: 6215.053 ms

```

Same plan (max_parallel_workers = 0). Less slowdown but still much slower than index scan
 ```
Limit  (cost=19763.14..19764.31 rows=10 width=12) (actual time=2648.495..2768.184 rows=10 loops=1)
   Buffers: shared hit=6634213 read=2
   InitPlan 1 (returns $0)
     ->  Limit  (cost=239.00..239.00 rows=1 width=40) (actual time=2.607..2.609 rows=1 loops=1)
           Buffers: shared hit=64
           ->  Sort  (cost=239.00..264.00 rows=10000 width=40) (actual time=2.607..2.608 rows=1 loops=1)
                 Sort Key: (random())
                 Sort Method: top-N heapsort  Memory: 25kB
                 Buffers: shared hit=64
                 ->  Seq Scan on emb_query  (cost=0.00..189.00 rows=10000 width=40) (actual time=0.021..1.
291 rows=10000 loops=1)
                       Buffers: shared hit=64
   ->  Gather Merge  (cost=19524.14..107030.25 rows=750000 width=12) (actual time=2648.493..2768.179 rows=
10 loops=1)
         Workers Planned: 2
         Params Evaluated: $0
         Workers Launched: 2
         Buffers: shared hit=6634213 read=2
         ->  Sort  (cost=18524.12..19461.62 rows=375000 width=12) (actual time=2643.702..2643.703 rows=7 l
oops=3)
               Sort Key: ((emb.vector <#> $0))
               Sort Method: top-N heapsort  Memory: 25kB
               Buffers: shared hit=6634149 read=2
               Worker 0:  Sort Method: top-N heapsort  Memory: 25kB
               Worker 1:  Sort Method: top-N heapsort  Memory: 25kB
               ->  Parallel Seq Scan on emb  (cost=0.00..10420.50 rows=375000 width=12) (actual time=0.301
..2601.952 rows=300000 loops=3)
                     Buffers: shared hit=6634075 read=2
 Planning:
   Buffers: shared hit=113
 Planning Time: 0.538 ms
 Execution Time: 2768.240 ms
```

with **enable_seqscan = off**:
```
 Limit  (cost=10000090607.85..10000090624.93 rows=10 width=12) (actual time=473.402..473.660 rows=10 loops
=1)
   Buffers: shared hit=2206 read=20567
   InitPlan 1 (returns $0)
     ->  Limit  (cost=10000000239.00..10000000239.00 rows=1 width=40) (actual time=2.642..2.644 rows=1 loo
ps=1)
           Buffers: shared hit=64
           ->  Sort  (cost=10000000239.00..10000000264.00 rows=10000 width=40) (actual time=2.642..2.643 r
ows=1 loops=1)
                 Sort Key: (random())
                 Sort Method: top-N heapsort  Memory: 25kB
                 Buffers: shared hit=64
                 ->  Seq Scan on emb_query  (cost=10000000000.00..10000000189.00 rows=10000 width=40) (act
ual time=0.020..1.300 rows=10000 loops=1)
                       Buffers: shared hit=64
   ->  Index Scan using emb_idx on emb  (cost=90368.85..1627254.85 rows=900000 width=12) (actual time=473.
401..473.655 rows=10 loops=1)
         Order By: (vector <#> $0)
         Buffers: shared hit=2206 read=20567
 Planning:
   Buffers: shared hit=113
 Planning Time: 0.499 ms
 Execution Time: 473.740 ms
(18 rows)
```
```
\d+ emb
                                                     Unlogged table "public.emb"
 Column |     Type     | Collation | Nullable |             Default             | Storage  | Compression |
 Stats target | Description
--------+--------------+-----------+----------+---------------------------------+----------+-------------+
--------------+-------------
 id     | integer      |           | not null | nextval('emb_id_seq'::regclass) | plain    |             |
              |
 vector | vector(1536) |           |          |                                 | extended |             |
              |
Indexes:
    "emb_pkey" PRIMARY KEY, btree (id)
    "emb_idx" ivfflat (vector vector_ip_ops) WITH (lists='2000')
```

**Comment by david-gang** [comment]
If you are using pgvector < 0.4.3 it could be connected to https://github.com/pgvector/pgvector/pull/133
which version do you use?

**Comment by pashkinelfe** [comment]
Yes, that's fixed in 0.4.3
